<a href="https://colab.research.google.com/github/MinervaNunez/Proyecto-de-Python-BEDU/blob/main/bedu_proyecto_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto de Python Equipo 8

>   *   Nicky García Fierros (nickygarcia@ciencias.unam.mx)
>   *   Kevin Martin Rivera Castro (kevinrc.unam@gmail.com)
>   *   Jesús Manuel Jimenez Cardoza (chucho003@hotmail.com)
>   *   Minerva Estefanía Núñez Manjarrez (estefania.nun.man@gmail.com)
>   *   Luis Mateo Patricio Pineda (mateopa@gmail.com)

---

## Librerías necesarias

---

In [91]:
import pandas as pd
import numpy as np
import requests

---

## Colección de Datos

---

### Lectura de datos

Leemos el archivo `.csv` a partir de la URL. En caso de error de conexión por falla del servidor, leemos el dataset directamente de nuestro repostitorio de GitHub.

In [92]:
df = pd.read_csv('https://raw.githubusercontent.com/MinervaNunez/Proyecto-de-Python-BEDU/main/RNPEDFC.csv', encoding='ISO-8859-1')

In [93]:
df

Fecha en que se le vio por ultima vez  \
0                                2012-03-18   
1                                2007-02-05   
2                                2009-03-14   
3                                2006-12-23   
4                                2008-01-11   
...                                     ...   
36260                            2018-04-26   
36261                            2018-01-08   
36262                            2015-02-02   
36263                            2017-08-22   
36264                            2018-04-11   

      Hora en que se le vio por ultima vez  \
0                                 05:00:00   
1                                 02:00:26   
2                                 08:30:54   
3                                 08:00:28   
4                                 12:00:17   
...                                    ...   
36260                             08:00:00   
36261                             11:00:00   
36262                             07:00:00   
36263                             04:00:00   
36264                             02:00:00   

      Pais en que se le vio por ultima vez  \
0                                   MEXICO   
1                                   MEXICO   
2                                   MEXICO   
3                                   MEXICO   
4                                   MEXICO   
...                                    ...   
36260                               MEXICO   
36261                               MEXICO   
36262                               MEXICO   
36263                               MEXICO   
36264                               MEXICO   

      Entidad en que se le vio por ultima vez  \
0                                      SONORA   
1                            ESTADO DE MEXICO   
2                            ESTADO DE MEXICO   
3                            ESTADO DE MEXICO   
4                            ESTADO DE MEXICO   
...                                       ...   
36260                                  PUEBLA   
36261                                  COLIMA   
36262                                  PUEBLA   
36263                                  COLIMA   
36264                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
0                                       NAVOJOA   
1                                VALLE DE BRAVO   
2                                        CHALCO   
3                           TLALNEPANTLA DE BAZ   
4                                        TOLUCA   
...                                         ...   
36260                                    PUEBLA   
36261                          VILLA DE ALVAREZ   
36262                                     CHILA   
36263                          VILLA DE ALVAREZ   
36264                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad Estatura  \
0                                       NAVOJOA     MEXICANA      1.1   
1                                VALLE DE BRAVO     MEXICANA      1.1   
2                                        CHALCO     MEXICANA      1.1   
3                           TLALNEPANTLA DE BAZ     MEXICANA      1.1   
4                                        TOLUCA     MEXICANA      1.1   
...                                         ...          ...      ...   
36260                                    PUEBLA     MEXICANA      1.6   
36261                          VILLA DE ALVAREZ     MEXICANA      1.7   
36262                      CHILAS DE LAS FLORES     MEXICANA     1.77   
36263                          VILLA DE ALVAREZ     MEXICANA     1.76   
36264                  SAN NICOLAS DE LOS GARZA     MEXICANA      1.6   

      Complexion    Sexo Edad  \
0        DELGADA   MUJER    7   
1        DELGADA   MUJER    4   
2        DELGADA  HOMBRE    6   
3        DELGADA  HOMBRE   10   
4        DELGADA  HOMBRE    4   
...          ...     ...  ...   
36260    DELGADA  HOMBRE 

##  Limpieza de datos

### Busqueda de NaNs

In [94]:
df.isna().sum()

Fecha en que se le vio por ultima vez        0
Hora en que se le vio por ultima vez         0
Pais en que se le vio por ultima vez         0
Entidad en que se le vio por ultima vez      0
Municipio en que se le vio por ultima vez    0
Localidad en que se le vio por ultima vez    0
Nacionalidad                                 0
Estatura                                     0
Complexion                                   0
Sexo                                         0
Edad                                         0
Descripcion de senas particulares            0
Etnia                                        0
Discapacidad                                 0
Dependencia que envio la informacion         0
dtype: int64

> Como nos percatamos en el análisis exploratorio, no contamos con valores faltantes indicados como `NaN`, sin embargo, identificamos que en nuestro dataset la categoría `NO ESPECIFICADO` y sus variaciones en cuanto a su escritura (`No Especificado`,  `no ESPECIFICADO`) fungen como registros de valores faltantes.

Vamos a realizar el conteo por columna de las ocurrencias del valor `NO ESPECIFICADO` (y sus variaciones) indicando el porcentaje que representa el número de faltantes respecto al total de registros en nuestro `DataFrame`. 

In [95]:
print(f'{("Columna"):42} =NO ESPECIFICADO=  Porcentaje\n')
for col in df.columns:
    total = (df[col] == "NO ESPECIFICADO").sum()+(df[col] == "No Especificado").sum()+(df[col] == "no ESPECIFICADO").sum()
    print(f'{(col):42} | {(total):16} {(round((total*100)/df.shape[0], 2)):9}%')

Columna                                    =NO ESPECIFICADO=  Porcentaje

Fecha en que se le vio por ultima vez      |              338      0.93%
Hora en que se le vio por ultima vez       |               16      0.04%
Pais en que se le vio por ultima vez       |                0       0.0%
Entidad en que se le vio por ultima vez    |               29      0.08%
Municipio en que se le vio por ultima vez  |              668      1.84%
Localidad en que se le vio por ultima vez  |             3413      9.41%
Nacionalidad                               |             2040      5.63%
Estatura                                   |            11008     30.35%
Complexion                                 |            11295     31.15%
Sexo                                       |                0       0.0%
Edad                                       |             3156       8.7%
Descripcion de senas particulares          |            18914     52.15%
Etnia                                      |      

> Tenemos columnas con nula presencia de faltantes, poca presencia y algunas columnas en donde la cantidad de faltantes es considerable. Pasemos a **tomar decisiones** sobre como trabajaremos con los valores faltantes en cada columna.

Primero crearemos una **copia de nuestro `DataFrame`** original para conservarlo.

In [96]:
df_original = df.copy()

Ahora transformemos todos nuestros valores de `NO ESPECIFICADO` (y sus variaciones) a `NaN`.

In [97]:
df = df.replace('NO ESPECIFICADO',np.NaN)
df = df.replace('No Especificado',np.NaN)
df = df.replace('no ESPECIFICADO',np.NaN)

In [98]:
df.isna().sum()

Fecha en que se le vio por ultima vez          338
Hora en que se le vio por ultima vez            16
Pais en que se le vio por ultima vez             0
Entidad en que se le vio por ultima vez         29
Municipio en que se le vio por ultima vez      668
Localidad en que se le vio por ultima vez     3413
Nacionalidad                                  2040
Estatura                                     11008
Complexion                                   11295
Sexo                                             0
Edad                                          3156
Descripcion de senas particulares            18914
Etnia                                        36133
Discapacidad                                  1047
Dependencia que envio la informacion             0
dtype: int64

> Podemos ver que el conteo de faltantes coincide con el que realizamos previamente.

### Limpiando NaNs

**Fecha en que se le vio por ultima vez**

La variable que registra la fecha en la que se vio por última vez a una persona es una variable muy importante en nuestro conjunto de datos y en el contexto del problema. Rellenar esta variable con algún tipo de dato puede hacer que el valor asignado pierda sentido o generar información errónea acerca del problema. Por esto eliminaremos los registros que contengan valores faltantes en esta columna.

In [99]:
df = df.dropna(subset=['Fecha en que se le vio por ultima vez'])

In [100]:
df.isna().sum()

Fecha en que se le vio por ultima vez            0
Hora en que se le vio por ultima vez             0
Pais en que se le vio por ultima vez             0
Entidad en que se le vio por ultima vez         26
Municipio en que se le vio por ultima vez      635
Localidad en que se le vio por ultima vez     3345
Nacionalidad                                  2028
Estatura                                     10784
Complexion                                   11157
Sexo                                             0
Edad                                          3088
Descripcion de senas particulares            18704
Etnia                                        35795
Discapacidad                                  1043
Dependencia que envio la informacion             0
dtype: int64

> Observamos que con las filas eliminadas a partir de los faltantes en las fechas, también eliminamos los registros faltantes en `Hora en que se le vio por última vez`, esto no representa una gran repercusión, pues al igual que la columna de fechas, los registros de horas eran datos importantes que no nos convenía rellenar y manipular.

**Pais en que se le vio por ultima vez**

Esta columna tiene el mismo valor (`MEXICO`) para todos los registros de la lista, por el contexto del problema podemos inferir el porqué de este valor. Podemos eliminar esta columna, pues no representa información relevante.

In [101]:
df = df.drop(columns=['Pais en que se le vio por ultima vez'])

**Etnia**

La columna `Etnia` tiene 36133 registros con el valor “no especificado”, es una cantidad muy grande y posiblemente no nos aporte mucho para un análisis. 

In [102]:
df = df.drop(columns=['Etnia'])

In [103]:
df.columns

Index(['Fecha en que se le vio por ultima vez',
       'Hora en que se le vio por ultima vez',
       'Entidad en que se le vio por ultima vez',
       'Municipio en que se le vio por ultima vez',
       'Localidad en que se le vio por ultima vez', 'Nacionalidad', 'Estatura',
       'Complexion', 'Sexo', 'Edad', 'Descripcion de senas particulares',
       'Discapacidad', 'Dependencia que envio la informacion'],
      dtype='object')

> Las columnas han sido eliminadas correctamente de nuestro `DataFrame`.

Remplazamos los valores faltantes por la media en el caso de columnas numéricas que son relevantes para nuestro análisis. No se eliminan porque reducirían bastante el `DataFrame`, y no se remplazan por cero porque sería un valor con un significado incoherente (altura cero es algo que no existe).

**Estatura**

Esta es nuestra variable objetivo, por el momento conservaremos los valores faltantes pero al realizar nuestro modelo nos centraremos unicamente en los valores con información.

Adelantaremos el proceso de transformación en esta columna para facilitar la imputación. Exploremos nuevamente los valores que toma esta variable.

In [104]:
df['Estatura'].value_counts()

1.7     4353
1.6     3314
1.65    2731
1.75    1959
1.8     1467
        ... 
0.45       1
0.51       1
0.57       1
1.33       1
90         1
Name: Estatura, Length: 152, dtype: int64

> La mayoría de los valores se encuentran en un rango de 0 a 2.04, esto nos dice que se maneja la medición en metros. Sin embargo, encontramos valores superiores a 50, es muy probable que estos estén registrados en centímetros. 

Vamos a convertir nuestros valores a tipo flotante y manejarlos en metros, haciendo la conversión respectiva.

In [105]:
df['Estatura'] = df['Estatura'].astype(float)
df['Estatura'] = df['Estatura'].apply(lambda x: x/100 if (x >= 3) else x)

> Ahora todas nuestras estaturas están representadas en metros.

In [106]:
aux = df[~df['Estatura'].isnull()].copy()

In [107]:
aux.isna().mean()

Fecha en que se le vio por ultima vez        0.000000
Hora en que se le vio por ultima vez         0.000000
Entidad en que se le vio por ultima vez      0.000795
Municipio en que se le vio por ultima vez    0.011295
Localidad en que se le vio por ultima vez    0.062443
Nacionalidad                                 0.041204
Estatura                                     0.000000
Complexion                                   0.083522
Sexo                                         0.000000
Edad                                         0.044386
Descripcion de senas particulares            0.396055
Discapacidad                                 0.019290
Dependencia que envio la informacion         0.000000
dtype: float64

**Edad**

Aplicaremos una transformación similar en nuestra columna de edades, convirtiendo nuestros valores a enteros. 

Pasemos a la imputación usando la mediana.

In [108]:
edad_mediana = df['Edad'].median()
edad_mediana

28.0

In [109]:
df['Edad'] = df['Edad'].fillna(edad_mediana)

In [110]:
df['Edad'].median().astype(int)

28

In [111]:
df["Edad"]= df["Edad"].astype(int)

> Ahora todas nuestras edades están representadas en enteros.

In [112]:
df.isna().sum()

Fecha en que se le vio por ultima vez            0
Hora en que se le vio por ultima vez             0
Entidad en que se le vio por ultima vez         26
Municipio en que se le vio por ultima vez      635
Localidad en que se le vio por ultima vez     3345
Nacionalidad                                  2028
Estatura                                     10784
Complexion                                   11157
Sexo                                             0
Edad                                             0
Descripcion de senas particulares            18704
Discapacidad                                  1043
Dependencia que envio la informacion             0
dtype: int64

> Nuestros valores faltantes en `Edad` han sido remplazados por la mediana, y la mediana con los valores remplazados sigue siendo igual a 28.

En las columnas donde los valores faltantes siguen presentes, vale la pena conservar la categoría de `NO ESPECIFICADO`. Regresaremos los valores `NaN` al valor `NO ESPECIFICADO` (esta vez sin ninguna variación) en las columnas restantes, esas columnas son:

*   `Entidad en que se le vio por ultima vez`
*   `Municipio en que se le vio por ultima vez`
*   `Localidad en que se le vio por ultima vez`
*   `Nacionalidad`
*   `Complexion`
*   `Descripcion de senas particulares`
*   `Discapacidad`

In [113]:
df.loc[:, df.columns != 'Estatura'] = df.loc[:, df.columns != 'Estatura'].fillna('NO ESPECIFICADO')

In [114]:
df.isna().sum()

Fecha en que se le vio por ultima vez            0
Hora en que se le vio por ultima vez             0
Entidad en que se le vio por ultima vez          0
Municipio en que se le vio por ultima vez        0
Localidad en que se le vio por ultima vez        0
Nacionalidad                                     0
Estatura                                     10784
Complexion                                       0
Sexo                                             0
Edad                                             0
Descripcion de senas particulares                0
Discapacidad                                     0
Dependencia que envio la informacion             0
dtype: int64

> Hemos limpiado nuestro 'DataFrame' de datos faltantes a expeción de la estatura (variable objetivo).

### Reindexando

In [115]:
df.tail(2)

Fecha en que se le vio por ultima vez  \
36263                            2017-08-22   
36264                            2018-04-11   

      Hora en que se le vio por ultima vez  \
36263                             04:00:00   
36264                             02:00:00   

      Entidad en que se le vio por ultima vez  \
36263                                  COLIMA   
36264                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
36263                          VILLA DE ALVAREZ   
36264                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad  Estatura  \
36263                          VILLA DE ALVAREZ     MEXICANA      1.76   
36264                  SAN NICOLAS DE LOS GARZA     MEXICANA      1.60   

      Complexion    Sexo  Edad  \
36263    ROBUSTA  HOMBRE    41   
36264    DELGADA   MUJER    14   

                       Descripcion de senas particulares Discapacidad  \
36263  TATUAJE EN EL BRAZO IZQUIERDO CON LA IMAGEN DE...      NINGUNO   
36264                                    NO ESPECIFICADO      NINGUNO   

      Dependencia que envio la informacion  
36263                         PGJ - COLIMA  
36264                     PGJ - NUEVO LEON

> Podemos ver que el número de filas ya no coincide con el índice del último registro. 

Vamos a reindexar y después establecer nuestros índices empezando desde 1.

In [116]:
df = df.reset_index(drop=True)

df.index = df.index + 1

In [117]:
df

Fecha en que se le vio por ultima vez  \
1                                2012-03-18   
2                                2007-02-05   
3                                2009-03-14   
4                                2006-12-23   
5                                2008-01-11   
...                                     ...   
35923                            2018-04-26   
35924                            2018-01-08   
35925                            2015-02-02   
35926                            2017-08-22   
35927                            2018-04-11   

      Hora en que se le vio por ultima vez  \
1                                 05:00:00   
2                                 02:00:26   
3                                 08:30:54   
4                                 08:00:28   
5                                 12:00:17   
...                                    ...   
35923                             08:00:00   
35924                             11:00:00   
35925                             07:00:00   
35926                             04:00:00   
35927                             02:00:00   

      Entidad en que se le vio por ultima vez  \
1                                      SONORA   
2                            ESTADO DE MEXICO   
3                            ESTADO DE MEXICO   
4                            ESTADO DE MEXICO   
5                            ESTADO DE MEXICO   
...                                       ...   
35923                                  PUEBLA   
35924                                  COLIMA   
35925                                  PUEBLA   
35926                                  COLIMA   
35927                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
1                                       NAVOJOA   
2                                VALLE DE BRAVO   
3                                        CHALCO   
4                           TLALNEPANTLA DE BAZ   
5                                        TOLUCA   
...                                         ...   
35923                                    PUEBLA   
35924                          VILLA DE ALVAREZ   
35925                                     CHILA   
35926                          VILLA DE ALVAREZ   
35927                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad  Estatura  \
1                                       NAVOJOA     MEXICANA      1.10   
2                                VALLE DE BRAVO     MEXICANA      1.10   
3                                        CHALCO     MEXICANA      1.10   
4                           TLALNEPANTLA DE BAZ     MEXICANA      1.10   
5                                        TOLUCA     MEXICANA      1.10   
...                                         ...          ...       ...   
35923                                    PUEBLA     MEXICANA      1.60   
35924                          VILLA DE ALVAREZ     MEXICANA      1.70   
35925                      CHILAS DE LAS FLORES     MEXICANA      1.77   
35926                          VILLA DE ALVAREZ     MEXICANA      1.76   
35927                  SAN NICOLAS DE LOS GARZA     MEXICANA      1.60   

      Complexion    Sexo  Edad  \
1        DELGADA   MUJER     7   
2        DELGADA   MUJER     4   
3        DELGADA  HOMBRE     6   
4        DELGADA  HOMBRE    10   
5        DELGADA  HOMBRE     4   
...          ...     ...   ...   
35923    DELGADA  HOMBRE    22   
35924    ROBUSTA   MUJER    27   
35925    MEDIANA   MUJER    86   
35926    ROBUSTA  HOMBRE    41   
35927    DELGADA   MUJER    14   

                       Descripcion de senas particulares Discapacidad  \
1      TIENE UNA CICATRIZ EN LA RODILLA IZQUIERDA POR...      NINGUNO   
2                                        NO ESPECIFICADO      NINGUNO   
3      MANCHAS, PARTE SUPERIOR DE PIERNA DERECHA, CIC...      NINGUNO   
4      LUNARES, A LA ALTURA DE LA CINTURAFORMA DE MAN...      NINGUNO   
5                      

> Ahora el número de filas ya no coincide con el índice del último registro y la indexación comienza desde 1. 

### Renombrando columnas

In [118]:
df.columns

Index(['Fecha en que se le vio por ultima vez',
       'Hora en que se le vio por ultima vez',
       'Entidad en que se le vio por ultima vez',
       'Municipio en que se le vio por ultima vez',
       'Localidad en que se le vio por ultima vez', 'Nacionalidad', 'Estatura',
       'Complexion', 'Sexo', 'Edad', 'Descripcion de senas particulares',
       'Discapacidad', 'Dependencia que envio la informacion'],
      dtype='object')

> Observamos el nombre de las columnas, podemos estandarizar los nombres para cumplir la convención snake case.

Construimos un diccionario con el mapeo de los nombres de las columnas esperados.

In [119]:
column_name_mapping = {
    'Fecha en que se le vio por ultima vez': 'fecha',
    'Hora en que se le vio por ultima vez': 'hora',
    'Entidad en que se le vio por ultima vez': 'entidad',
    'Municipio en que se le vio por ultima vez': 'municipio',
    'Localidad en que se le vio por ultima vez': 'localidad',
    'Nacionalidad': 'nacionalidad',
    'Estatura': 'estatura',
    'Complexion': 'complexion',
    'Sexo': 'sexo',
    'Edad': 'edad',
    'Descripcion de senas particulares': 'senas_particulares',
    'Discapacidad': 'discapacidad',
    'Dependencia que envio la informacion': 'dependecia_origen'
    
}

df = df.rename(columns=column_name_mapping)
df

fecha      hora           entidad                 municipio  \
1      2012-03-18  05:00:00            SONORA                   NAVOJOA   
2      2007-02-05  02:00:26  ESTADO DE MEXICO            VALLE DE BRAVO   
3      2009-03-14  08:30:54  ESTADO DE MEXICO                    CHALCO   
4      2006-12-23  08:00:28  ESTADO DE MEXICO       TLALNEPANTLA DE BAZ   
5      2008-01-11  12:00:17  ESTADO DE MEXICO                    TOLUCA   
...           ...       ...               ...                       ...   
35923  2018-04-26  08:00:00            PUEBLA                    PUEBLA   
35924  2018-01-08  11:00:00            COLIMA          VILLA DE ALVAREZ   
35925  2015-02-02  07:00:00            PUEBLA                     CHILA   
35926  2017-08-22  04:00:00            COLIMA          VILLA DE ALVAREZ   
35927  2018-04-11  02:00:00        NUEVO LEON  SAN NICOLAS DE LOS GARZA   

                      localidad nacionalidad  estatura complexion    sexo  \
1                       NAVOJOA     MEXICANA      1.10    DELGADA   MUJER   
2                VALLE DE BRAVO     MEXICANA      1.10    DELGADA   MUJER   
3                        CHALCO     MEXICANA      1.10    DELGADA  HOMBRE   
4           TLALNEPANTLA DE BAZ     MEXICANA      1.10    DELGADA  HOMBRE   
5                        TOLUCA     MEXICANA      1.10    DELGADA  HOMBRE   
...                         ...          ...       ...        ...     ...   
35923                    PUEBLA     MEXICANA      1.60    DELGADA  HOMBRE   
35924          VILLA DE ALVAREZ     MEXICANA      1.70    ROBUSTA   MUJER   
35925      CHILAS DE LAS FLORES     MEXICANA      1.77    MEDIANA   MUJER   
35926          VILLA DE ALVAREZ     MEXICANA      1.76    ROBUSTA  HOMBRE   
35927  SAN NICOLAS DE LOS GARZA     MEXICANA      1.60    DELGADA   MUJER   

       edad                                 senas_particulares discapacidad  \
1         7  TIENE UNA CICATRIZ EN LA RODILLA IZQUIERDA POR...      NINGUNO   
2         4                                    NO ESPECIFICADO      NINGUNO   
3         6  MANCHAS, PARTE SUPERIOR DE PIERNA DERECHA, CIC...      NINGUNO   
4        10  LUNARES, A LA ALTURA DE LA CINTURAFORMA DE MAN...      NINGUNO   
5         4                                    NO ESPECIFICADO      NINGUNO   
...     ...                                                ...          ...   
35923    22  LUNAR EN LA MEJILLA Y TATUAJE EN EL CUELLO EN ...      NINGUNO   
35924    27  EL OJO IZQUIERDO TIENE UN ENFERMEDAD COMO TIC,...      NINGUNO   
35925    86                                    NO ESPECIFICADO      NINGUNO   
35926    41  TATUAJE EN EL BRAZO IZQUIERDO CON LA IMAGEN DE...      NINGUNO   
35927    14                                    NO ESPECIFICADO      NINGUNO   

            dependecia_origen  
1                FGE - SONORA  
2      FGJ - ESTADO DE MEXICO  
3      FGJ - ESTADO DE MEXICO  
4      FGJ - ESTADO DE MEXICO  
5      FGJ - ESTADO DE MEXICO  
...                       ...  
35923            FGE - PUEBLA  
35924            PGJ - COLIMA  
35925            FGE - PUEBLA  
35926            PGJ - COLIMA  
35927        PGJ - NUEVO LEON  

[35927 rows x 13 columns]

> Nuestros nombres se encuentran estandarizados y cumplen la convención **snake case**.

---

## Transformación, filtración y ordenamiento de datos

---

### Casting

Observemos nuevamente nuestras columnas y su tipo de dato.

In [120]:
df.dtypes

fecha                  object
hora                   object
entidad                object
municipio              object
localidad              object
nacionalidad           object
estatura              float64
complexion             object
sexo                   object
edad                    int64
senas_particulares     object
discapacidad           object
dependecia_origen      object
dtype: object

> Ya realizamos el casting correspondiente en nuestras variables numéricas `estatura` y `edad` pero aún tenemos dos columnas que tienen un tipo de dato incorrecto. 

Usaremos el método `astype` y un diccionario de conversión para cambiar el tipo de datos de las columnas `fecha` y `hora`.

In [121]:
diccionario_de_conversion = {
    'fecha': 'datetime64',
    'hora': 'datetime64'
}

In [122]:
df = df.astype(diccionario_de_conversion)

In [123]:
df['fecha']

1       2012-03-18
2       2007-02-05
3       2009-03-14
4       2006-12-23
5       2008-01-11
           ...    
35923   2018-04-26
35924   2018-01-08
35925   2015-02-02
35926   2017-08-22
35927   2018-04-11
Name: fecha, Length: 35927, dtype: datetime64[ns]

In [124]:
df['hora']

1       2022-12-09 05:00:00
2       2022-12-09 02:00:26
3       2022-12-09 08:30:54
4       2022-12-09 08:00:28
5       2022-12-09 12:00:17
                ...        
35923   2022-12-09 08:00:00
35924   2022-12-09 11:00:00
35925   2022-12-09 07:00:00
35926   2022-12-09 04:00:00
35927   2022-12-09 02:00:00
Name: hora, Length: 35927, dtype: datetime64[ns]

> Las variables fecha y hora ya son de tipo `datetime64`, sin embargo, a los valores de hora se les ha agregado la fecha actual, debemos corregir esto.

Extraemos únicamente el registro del tiempo de la columna `hora`.

In [125]:
df['hora'] = df['hora'].dt.time
df['hora']

1        05:00:00
2        02:00:26
3        08:30:54
4        08:00:28
5        12:00:17
           ...   
35923    08:00:00
35924    11:00:00
35925    07:00:00
35926    04:00:00
35927    02:00:00
Name: hora, Length: 35927, dtype: object

In [126]:
df.dtypes

fecha                 datetime64[ns]
hora                          object
entidad                       object
municipio                     object
localidad                     object
nacionalidad                  object
estatura                     float64
complexion                    object
sexo                          object
edad                           int64
senas_particulares            object
discapacidad                  object
dependecia_origen             object
dtype: object

> Vemos que el formato de `fecha` ahora aparece como `datetime64[ns]`, en el caso de la columna `hora` lo sigue mostrando como object porque es una extracción de  `datetime64[ns]`.

### Manipulación de Strings

In [127]:
df.head(3)

fecha      hora           entidad       municipio       localidad  \
1 2012-03-18  05:00:00            SONORA         NAVOJOA         NAVOJOA   
2 2007-02-05  02:00:26  ESTADO DE MEXICO  VALLE DE BRAVO  VALLE DE BRAVO   
3 2009-03-14  08:30:54  ESTADO DE MEXICO          CHALCO          CHALCO   

  nacionalidad  estatura complexion    sexo  edad  \
1     MEXICANA       1.1    DELGADA   MUJER     7   
2     MEXICANA       1.1    DELGADA   MUJER     4   
3     MEXICANA       1.1    DELGADA  HOMBRE     6   

                                  senas_particulares discapacidad  \
1  TIENE UNA CICATRIZ EN LA RODILLA IZQUIERDA POR...      NINGUNO   
2                                    NO ESPECIFICADO      NINGUNO   
3  MANCHAS, PARTE SUPERIOR DE PIERNA DERECHA, CIC...      NINGUNO   

        dependecia_origen  
1            FGE - SONORA  
2  FGJ - ESTADO DE MEXICO  
3  FGJ - ESTADO DE MEXICO

Los textos de nuestras variables de tipo `string` están en mayúsculas. Vamos a modificar el patrón iniciando con mayúscula la primera letra de cada palabra.

In [128]:
columnas = ['entidad', 'municipio', 'localidad', 'nacionalidad', 'complexion', 'sexo', 'senas_particulares', 'discapacidad']

for col in columnas:
    df[col] = df[col].str.title()

En el caso de los valores de dependencia origen nos encontramos con las siglas de las dependencias, por lo tanto, debemos mantenerlas de esta forma. Podemos extraer las letras correspondientes a las siglas y aplicar `.str.title()` únicamente en el resto del texto.

In [129]:
df['dependecia_origen'] = df['dependecia_origen'].str[:6] + df['dependecia_origen'].str[6:].str.title()

In [130]:
df

fecha      hora           entidad                 municipio  \
1     2012-03-18  05:00:00            Sonora                   Navojoa   
2     2007-02-05  02:00:26  Estado De Mexico            Valle De Bravo   
3     2009-03-14  08:30:54  Estado De Mexico                    Chalco   
4     2006-12-23  08:00:28  Estado De Mexico       Tlalnepantla De Baz   
5     2008-01-11  12:00:17  Estado De Mexico                    Toluca   
...          ...       ...               ...                       ...   
35923 2018-04-26  08:00:00            Puebla                    Puebla   
35924 2018-01-08  11:00:00            Colima          Villa De Alvarez   
35925 2015-02-02  07:00:00            Puebla                     Chila   
35926 2017-08-22  04:00:00            Colima          Villa De Alvarez   
35927 2018-04-11  02:00:00        Nuevo Leon  San Nicolas De Los Garza   

                      localidad nacionalidad  estatura complexion    sexo  \
1                       Navojoa     Mexicana      1.10    Delgada   Mujer   
2                Valle De Bravo     Mexicana      1.10    Delgada   Mujer   
3                        Chalco     Mexicana      1.10    Delgada  Hombre   
4           Tlalnepantla De Baz     Mexicana      1.10    Delgada  Hombre   
5                        Toluca     Mexicana      1.10    Delgada  Hombre   
...                         ...          ...       ...        ...     ...   
35923                    Puebla     Mexicana      1.60    Delgada  Hombre   
35924          Villa De Alvarez     Mexicana      1.70    Robusta   Mujer   
35925      Chilas De Las Flores     Mexicana      1.77    Mediana   Mujer   
35926          Villa De Alvarez     Mexicana      1.76    Robusta  Hombre   
35927  San Nicolas De Los Garza     Mexicana      1.60    Delgada   Mujer   

       edad                                 senas_particulares discapacidad  \
1         7  Tiene Una Cicatriz En La Rodilla Izquierda Por...      Ninguno   
2         4                                    No Especificado      Ninguno   
3         6  Manchas, Parte Superior De Pierna Derecha, Cic...      Ninguno   
4        10  Lunares, A La Altura De La Cinturaforma De Man...      Ninguno   
5         4                                    No Especificado      Ninguno   
...     ...                                                ...          ...   
35923    22  Lunar En La Mejilla Y Tatuaje En El Cuello En ...      Ninguno   
35924    27  El Ojo Izquierdo Tiene Un Enfermedad Como Tic,...      Ninguno   
35925    86                                    No Especificado      Ninguno   
35926    41  Tatuaje En El Brazo Izquierdo Con La Imagen De...      Ninguno   
35927    14                                    No Especificado      Ninguno   

            dependecia_origen  
1                FGE - Sonora  
2      FGJ - Estado De Mexico  
3      FGJ - Estado De Mexico  
4      FGJ - Estado De Mexico  
5      FGJ - Estado De Mexico  
...                       ...  
35923            FGE - Puebla  
35924            PGJ - Colima  
35925            FGE - Puebla  
35926            PGJ - Colima  
35927        PGJ - Nuevo Leon  

[35927 rows x 13 columns]

> Ahora todas nuestras columnas texto, tienen un formato uniforme. 

---

## Exportar DataFrame resultante a CSV

---

Antes de guardar nuestro `DataFrame` final vamos a reordenar nuestras columnas para que nuestros datos tengan un mejor orden.

In [132]:
df = df[['fecha', 'hora',
         'entidad', 'municipio', 'localidad', 'nacionalidad',
         'sexo', 'edad', 'estatura', 'complexion', 'senas_particulares','discapacidad', 'dependecia_origen']]
df

fecha      hora           entidad                 municipio  \
1     2012-03-18  05:00:00            Sonora                   Navojoa   
2     2007-02-05  02:00:26  Estado De Mexico            Valle De Bravo   
3     2009-03-14  08:30:54  Estado De Mexico                    Chalco   
4     2006-12-23  08:00:28  Estado De Mexico       Tlalnepantla De Baz   
5     2008-01-11  12:00:17  Estado De Mexico                    Toluca   
...          ...       ...               ...                       ...   
35923 2018-04-26  08:00:00            Puebla                    Puebla   
35924 2018-01-08  11:00:00            Colima          Villa De Alvarez   
35925 2015-02-02  07:00:00            Puebla                     Chila   
35926 2017-08-22  04:00:00            Colima          Villa De Alvarez   
35927 2018-04-11  02:00:00        Nuevo Leon  San Nicolas De Los Garza   

                      localidad nacionalidad    sexo  edad  estatura  \
1                       Navojoa     Mexicana   Mujer     7      1.10   
2                Valle De Bravo     Mexicana   Mujer     4      1.10   
3                        Chalco     Mexicana  Hombre     6      1.10   
4           Tlalnepantla De Baz     Mexicana  Hombre    10      1.10   
5                        Toluca     Mexicana  Hombre     4      1.10   
...                         ...          ...     ...   ...       ...   
35923                    Puebla     Mexicana  Hombre    22      1.60   
35924          Villa De Alvarez     Mexicana   Mujer    27      1.70   
35925      Chilas De Las Flores     Mexicana   Mujer    86      1.77   
35926          Villa De Alvarez     Mexicana  Hombre    41      1.76   
35927  San Nicolas De Los Garza     Mexicana   Mujer    14      1.60   

      complexion                                 senas_particulares  \
1        Delgada  Tiene Una Cicatriz En La Rodilla Izquierda Por...   
2        Delgada                                    No Especificado   
3        Delgada  Manchas, Parte Superior De Pierna Derecha, Cic...   
4        Delgada  Lunares, A La Altura De La Cinturaforma De Man...   
5        Delgada                                    No Especificado   
...          ...                                                ...   
35923    Delgada  Lunar En La Mejilla Y Tatuaje En El Cuello En ...   
35924    Robusta  El Ojo Izquierdo Tiene Un Enfermedad Como Tic,...   
35925    Mediana                                    No Especificado   
35926    Robusta  Tatuaje En El Brazo Izquierdo Con La Imagen De...   
35927    Delgada                                    No Especificado   

      discapacidad       dependecia_origen  
1          Ninguno            FGE - Sonora  
2          Ninguno  FGJ - Estado De Mexico  
3          Ninguno  FGJ - Estado De Mexico  
4          Ninguno  FGJ - Estado De Mexico  
5          Ninguno  FGJ - Estado De Mexico  
...            ...                     ...  
35923      Ninguno            FGE - Puebla  
35924      Ninguno            PGJ - Colima  
35925      Ninguno            FGE - Puebla  
35926      Ninguno            PGJ - Colima  
35927      Ninguno        PGJ - Nuevo Leon  

[35927 rows x 13 columns]

Finalmente, el `DataFrame` resultante se puede exportar como archivo `.csv`.

In [133]:
df.to_csv('RNPEDFC_Final.csv', index=False, encoding='ISO-8859-1')